Add e2e tests

samreaves opened the pull request:
#### To run end to end tests: 
1. Start up the webpack dev server with either gulp serve or gulp watch.
2. Run "npm run e2e" to get automated tests. By just running "protractor" in the root directory, you will see test logs with greater verbosity.


#### E2E Test Generator
I've added a component e2e test file generator, and updated the component module generator with a default route of the component name. In order for the component to be tested, two tasks must still be accomplished.

1. New component must be added to the components.js file.
2. There must be a link to the route somewhere in the app for ui-router to recognize it.

fesor commented on package.json:
```diff
@@ -44,6 +45,10 @@
   },
   "scripts": {
+    "preupdate-webdriver": "npm install",
+    "update-webdriver": "webdriver-manager update",
```
Maybe run it as `postinstall` script?

fesor commented on package.json:
```diff
@@ -44,6 +45,10 @@
   "scripts": {
+    "preupdate-webdriver": "npm install",
+    "update-webdriver": "webdriver-manager update",
+    "pree2e": "npm run update-webdriver",
```
I think that we should move this to gulpfile, since it have all needed info about server host and port.

fesor commented on package.json:
```diff
@@ -44,6 +45,10 @@
+    "preupdate-webdriver": "npm install",
+    "update-webdriver": "webdriver-manager update",
+    "pree2e": "npm run update-webdriver",
+    "e2e": "protractor protractor.conf.js",
```
I think that we should move this to gulpfile, since it have all needed info about server host and port.

BelfordZ commented on protractor.conf.js:
```diff
@@ -0,0 +1,30 @@
+	allScriptsTimeout: 30000,
+	onPrepare: function () {
+    require("babel-core/register")({ retainLines: true });
+    browser.ignoreSynchronization = true;
```
This is not best practice and should be avoided. If you want to ignore synchronization, do it on a test-by-test basis rather than app wide.

BelfordZ commented on generator/component/temp.e2e.js:
```diff
@@ -0,0 +1,16 @@
+	beforeEach(() => {
+
+		// Navigate to <%= name %> component
+    browser.get('http://localhost:3000/<%= name %>');
```
urls called with browser get are relative to the base (you dont need to put localhost:3000 there)

samreaves commented:
How about now, guys? 

Removed extraneous code, added environment variables, using relative paths, and the developer experience is so much better.

### gulp e2e
- bundles our source with webpack
- serves a webserver 
- updates selenium 
- runs selenium
- runs protractor (with verbosity this time!)
- exits process after protractor is finished

### npm e2e
- Runs gulp e2e

fesor commented on gulpfile.babel.js:
```diff
@@ -60,7 +61,31 @@ gulp.task('webpack', (cb) => {
+
+
+gulp.task('e2e', ['protractor'], function() {
+  process.exit();
```
This is maybe not a good idea, but we could store server instance in gulpfile scope and then just stop it. But i thinkg this is better than `process.exit`.

samreaves replied:
In the previous commit I have serve.exit() [from the browser-sync API](https://www.browsersync.io/docs/api/) instead, and it doesn't stop the server. Exiting the process seemed to be the only option.

fesor replied:
It doesn't stop gulp task maybe?

```
let stopServer;

// ...
gulp.task('server', function (callback) {
    const server = browsersync({...});

    stopServer = () => {
       server.exit();
       callback();
    };

});

// ...

stopServer();
```

samreaves replied:
I think you're right. However with that implementation, protractor never starts. Probably because serve task is expecting to only proceed after the callback?

fesor replied:
Yah, you are right... so we need to run callback after server is ready.

samreaves replied:
Unsure what I'm missing here. Everything finishes except for the process.

`
// Setting up the test task 
gulp.task('protractor', ['serve'], function(callback) {
    gulp
        .src(['example_spec.js'])
        .pipe(gulpProtractorAngular({
            'configFile': 'protractor.conf.js',
            'debug': false,
            'autoStartStopServer': true
        }))
        .on('error', function(e) {
            console.log(e);
        })
        .on('end', callback);
});


gulp.task('e2e', ['protractor'], function(callback) {
  //process.exit();
  stopServer();
  callback();
});



gulp.task('serve', function(callback) {
  const config = require('./webpack.dev.config');
  config.entry.app = [
    // this modules required to make HRM working
    // it responsible for all this webpack magic
    'webpack-hot-middleware/client?reload=true',
    // application entry point
    paths.entry
  ];

  var compiler = webpack(config);

  const server = serve.init({
    port: process.env.PORT || 3000,
    open: false,
    server: {baseDir: root},
    middleware: [
      historyApiFallback(),
      webpackDevMiddelware(compiler, {
        stats: {
          colors: colorsSupported,
          chunks: false,
          modules: false
        },
        publicPath: config.output.publicPath
      }),
      webpachHotMiddelware(compiler)
    ]
  });

  stopServer = function(callback) {
    server.exit();
  }

  callback();
});
`

pziegler commented:
update on this ? Ive used most of this stuff and I can get the protractor to run properly, but as you say, when used with gulp it doesnt want to exit..

fesor commented:
I currently working on getting rid of gulp completely. This will simplify stuff.

FarhadG commented:
Do we have a working version of the latest here? Curious...
